### ISSUE TITLE
Feature request for 'Should'

### ISSUE THREAD

**Issue by DarkLite1** [pester/Pester#1712]
## 1. General summary of the issue

Would it be possible for `Should` to produce a `boolean` value so it can be used within the script block of `-ParameterFilter ` available on `Mock`? See the example code below to clarify. 

```powershell
Describe '-ParameterFilter test' {
    BeforeAll {
        Function Test-Colors {
            Param (
                [String[]]$Colors
            )
            $Colors.Contains('Green')
        }

        Mock Test-Colors

        Test-Colors -Colors @('Green', 'Red')
    }
    It "desired functionality" {
        Should -invoke Test-Colors -Scope Describe -ParameterFilter {
            $Colors | Should -Be @('Green', 'Red')
        }
    }
    It "workaround" {
        Should -invoke Test-Colors -Scope Describe -ParameterFilter {
            $Colors | Should -Be @('Green', 'Red')
            $true
        }
    }
}
```
2. Possible solution

Maybe it would be great if a switch could be added like:

```powershell
@('apple', 'kiwi') | Should -Be -OutBoolean @('apple', 'kiwi')
```

**Comment by fflaten** [comment]
What are you trying to solve? Why can't you use `$Colors -contains 'Green' -and $Colors -contains 'Red'`?

**Comment by DarkLite1** [comment]
I could. there's nothing wrong wit that. But what if I wanted to check more than one item? Or to be exactly equal to another array? The test code would just become longer and longer or I would have to create my own helper function like AreArraysEqual as [someone else ](https://www.red-gate.com/simple-talk/sysadmin/powershell/practical-powershell-unit-testing-checking-program-flow/#fifth)did.

That's a bit redudant if Pester already has this capable with the [`should` assertions](https://pester.dev/docs/usage/assertions). . Then simply why not allow them to be used within `-ParameterFilter` that only requires a `boolean` result?

**Comment by fflaten** [comment]
I see. Bad example on my part, should've been

**Comment by nohwnd** [comment]
I think we can wrap the parameter filter evaluation in try catch, and consider the filter to be failed if there is an exception. Otherwise I am not sure how we would handle multiple Shoulds in the same filter, because they would just return $true / $false so all of them would run, and then we would check if there was at least some failure, but it would be very hard to see which one. Instead we can catch the error and write it to the Diagnostic log, when enabled.

**Comment by DarkLite1** [comment]
Yeah, I've noticed when combining the `should` test with  normal `x -eq x` the testresult becomes unreliable. 

Depending on if the line `$Colors[0] -eq 'wrong'` is before or after the should it might fail or succeed.

```powershell
        Should -invoke Test-Colors -Scope Describe -ParameterFilter {
            $Colors[0] -eq 'wrong'
            $Colors | Should -Be @('Green', 'Red')
            $true
        }
```

**Comment by johlju** [comment]
I like this idea, today for it to handle `-Times <number>` correctly when using a parameter filter one must do something this:

```powershell
Should -Invoke -CommandName Write-Verbose -ParameterFilter {
    # Only mock the command that output the string that should be tested.
    $correctMessage = $Message -match 'My Message'

    # Only test string if it is the correct verbose command
    if ($correctMessage)
    {
        $Message | Should -MatchExactly $RegEx1
        $Message | Should -MatchExactly $RegEx2
    }

    # Return wether the correct command was called or not.
    $correctMessage
} -Exactly -Times 1 -Scope It
```

In the above snippet it would not help much to have `OutBoolean` (or `PassThru`) because as @nohwnd said the first `Should` could return `$true` while the other could return `$false` and it shuld only be evaulated IF it is the correct call of (in this case) `Write-Verbose`.  

Would it work if there were no value of `$true` returned it is always assumed that it is `$false`, and always throw on a `Should` that fails evaluation. But then in the snippet above we only be able to get rid of `$correctMessage`:

```powershell
Should -Invoke -CommandName Write-Verbose -ParameterFilter {
    # Only mock the command that output the string that should be tested.
    $correctMessage = $Message -match 'My Message'

    # Only test string if it is the correct verbose command
    if ($correctMessage)
    {
        $Message | Should -MatchExactly $RegEx1 -PassThru # Or -OutBoolean
        $Message | Should -MatchExactly $RegEx2 -PassThru # Or -OutBoolean
    }
} -Exactly -Times 1 -Scope It
```

**Comment by nohwnd** [comment]
Closing this - it already works with both Should and the new Should-* assertions. Inside a -ParameterFilter a failing assertion throws and is treated as a non-match, so:

    -ParameterFilter { $Colors | Should -Be @('Green','Red'); $true }

gives you exactly the boolean behavior you want (the trailing $true is only reached when the assertion passes). Making Should itself return a bool would change the fail-fast semantics of every assertion, which I don't want to do. Thanks!
